service/worker/handler/api_replication_switch.go

```diff
@@ -21,59 +21,85 @@ import (
 	"encoding/json"
 	"errors"
 	"fmt"
+	"time"
 
 	"github.com/hibiken/asynq"
 	"github.com/rs/zerolog"
 
 	"github.com/clyso/chorus/pkg/dom"
+	"github.com/clyso/chorus/pkg/lock"
 	"github.com/clyso/chorus/pkg/log"
+	"github.com/clyso/chorus/pkg/policy"
 	"github.com/clyso/chorus/pkg/tasks"
 )
 
-func (s *svc) FinishReplicationSwitch(ctx context.Context, t *asynq.Task) error {
-	var p tasks.FinishReplicationSwitchPayload
+func (s *svc) HandleZeroDowntimeReplicationSwitch(ctx context.Context, t *asynq.Task) error {
+	var p tasks.ZeroDowntimeReplicationSwitchPayload
 	if err := json.Unmarshal(t.Payload(), &p); err != nil {
-		return fmt.Errorf("FinishReplicationSwitchPayload Unmarshal failed: %w: %w", err, asynq.SkipRetry)
+		return fmt.Errorf("ZeroDowntimeReplicationSwitchPayload Unmarshal failed: %w: %w", err, asynq.SkipRetry)
 	}
 	ctx = log.WithBucket(ctx, p.Bucket)
-	logger := zerolog.Ctx(ctx)
 
-	replSwitch, err := s.policySvc.GetReplicationSwitch(ctx, p.User, p.Bucket)
+	policyID := policy.ReplicationID{
+		User:     p.User,
+		Bucket:   p.Bucket,
+		From:     p.FromStorage,
+		To:       p.ToStorage,
+		ToBucket: p.ToBucket,
+	}
+
+	// acquire exclusive lock to switch task:
+	release, refresh, err := s.locker.Lock(ctx, lock.StringKey(policyID.String()))
 	if err != nil {
-		if errors.Is(err, dom.ErrNotFound) {
-			logger.Warn().Msg("unable to finish replication switch task: not found")
-			return nil
-		}
 		return err
 	}
-	if replSwitch.IsDone {
-		logger.Info().Msg("skip finish replication switch task: already done")
-		return nil
-	}
-
-	for oldFollower := range replSwitch.GetOldFollowers() {
-		oldRepl, err := s.policySvc.GetReplicationPolicyInfo(ctx, p.User, p.Bucket, replSwitch.OldMain, string(oldFollower), nil)
+	defer release()
+	return lock.WithRefresh(ctx, func() error {
```

Comment:
Can be beneficial to move `func` out for readability.

Reply:
thanks, done